<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="../css/text_page.css" rel="stylesheet" type="text/css" />
    <script src="https://kit.fontawesome.com/9533923917.js" crossorigin="anonymous"></script>

    <title>How to Upload Your Own Data | QuickTree</title>

    <link rel="apple-touch-icon" sizes="180x180" href="../favicon/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="../favicon/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="../favicon/favicon-16x16.png">
    <link rel="manifest" href="../favicon/site.webmanifest">
    <link rel="mask-icon" href="../favicon/safari-pinned-tab.svg" color="#679436">
    <meta name="msapplication-TileColor" content="#9bc04f">
    <meta name="theme-color" content="#ffffff">
    <link rel="shortcut icon" href="../favicon.ico">
</head>

<body>
    <a href="tree_drop.html" class="button">
        <button type="button" class="btn btn-primary"> <i class="fas fa-arrow-left"></i> Back </button>
    </a>
    <div>

        <h1><a href="../index.html">QuickTree</a></h1>
        <hr/>
        <h1>How to Upload Your Own Data</h1>
        <hr/>
        <p>
            These step-by-step instructions will make sure you know EXACTLY what to do if you want
            to use your own data. What are you waiting for? Let's make a tree!
        </p>

        <ol type="1">

            <li STYLE="font-weight: bold">
                Create the necessary files:
                <ol type="a">
                    <li STYLE="font-weight: normal">
                        Create a FASTA file to hold the DNA sequences.
                        <ul>
                            <li>
                                Name the file "speciesFASTA.txt" (replace "species" with the type of
                                species that will be displayed in the tree, such as "birds", "flowers", or
                                "mammals").
                            </li>
                            <li>This file will be used to generate the phylogenetic tree.</li>
                        </ul>
                    </li>
                    <li STYLE="font-weight: normal">
                        Create a JSON file to hold the phylogenetic tree.
                        <ul>
                            <li>Name the file "speciesTree.json"</li>
                        </ul>
                    </li>
                    <li STYLE="font-weight: normal">
                        Create a JSON file to hold the species data.
                        <ul>
                            <li>Name the file "speciesData.json"</li>
                            <li>This file will be used for the species information view.</li>
                        </ul>
                    </li>
                    <li STYLE="font-weight: normal">
                        Create a JSON file to hold the DNA comparisons.
                        <ul>
                            <li>Name the file "speciesComparisons.json"</li>
                            <li>This file will be used for the DNA comparison view.</li>
                        </ul>
                    </li>
                </ol>
            </li>

            <br/>

            <li STYLE="font-weight: bold">
                Collect the DNA sequences:
                <ol type="a">
                    <li STYLE="font-weight: normal">
                        If you're not sure where to start, pick a kingdom, phylum, class, order, or
                        family and browse the hierarchy in Wikipedia to get a feel for what species are
                        out there. (Not every species has been sequenced, but there are many that have been.)
                    </li>
                    <li STYLE="font-weight: normal">
                        For each desired species...
                        <ul>
                            <li>
                                Search GenBank for "[scientific name of species] mitochondrion". (This will work as long as
                                the organisms in the tree are eukaryotes.)
                            </li>
                            <li>
                                If there is a sequence for the complete mitochondrial genome of your species, select the "FASTA"
                                button of that search result. (Some species have many options to choose from, while others have none.)
                            </li>
                            <li>
                                Copy and paste the the lines of text into "speciesFASTA.txt".
                            </li>
                            <li>
                                Take note of the accession number, then replace the text on the first
                                line with ">Common Name of Species|(Scientific Name of Species)".
                            </li>
                            <li>
                                Leave an empty line between each DNA sequence.
                            </li>
                        </ul>
                    </li>
                </ol>
            </li>

            <br/>

            <li STYLE="font-weight: bold">
                Create the phylogenetic tree:
                <ol type="a">
                    <li STYLE="font-weight: normal">
                        In R, install the library <code>ape</code>.
                    </li>
                    <li STYLE="font-weight: normal">
                        Download the executable for <a href="https://www.drive5.com/muscle/downloads.htm">MUSCLE</a>
                        and/or <a href="http://www.clustal.org/download/current/">CLUSTAL</a> sequence alignment. Make
                        sure the directory containing the executables is added to your PATH environmental variable.
                    </li>
                    <li STYLE="font-weight: normal">
                        In R, type the following lines of code (replacing "species" where appropriate):
                        <pre>

library(ape)
speciesFASTA = "path/to/fasta/file/speciesFASTA.txt"
species = read.dna(speciesFASTA, format = "fasta")
speciesMUSCLE = muscle(species)  # or speciesCLUSTAL = clustal(species)
speciesTree = bionj(dist.dna(speciesMUSCLE))  # or speciesTree = bionj(dist.dna(speciesCLUSTAL))
plot(speciesTree)  # Just to make sure it looks how you want
write.tree(speciesTree, file = "speciesNEWICK.txt")
                        </pre>
                    </li>
                    <li STYLE="font-weight: normal">
                        If something doesn't look right, you can add or remove DNA sequences in
                        "speciesFASTA.txt" and run the lines of code again.
                    </li>
                    <li STYLE="font-weight: normal">
                        Open "speciesTree.json" and paste the following:
                        <pre>

[
  {
    "name": "species",
    "tree": ""
  }
]
                        </pre>
                    </li>
                    <li STYLE="font-weight: normal">
                        Copy the contents of "speciesNEWICK.txt" and paste inside the
                        quotes after <code>"tree":</code> in "speciesTree.json".
                    </li>
                </ol>
            </li>

            <br/>

            <li STYLE="font-weight: bold">
                Collect the species data:
                <ol type="a">
                    <li STYLE="font-weight: normal">
                        Open "speciesData.json" and paste the following (with one JSON object for each
                        species in your tree):
                        <pre>

[
  {
    "scientificName": "",
    "commonName": "",
    "accessionNumber": "",
    "mapLink": "",
    "imageUrl": "",
    "dnaSequences": [
      "",
      "",
      "",
      ""
    ]
  },
  {
    "scientificName": "",
    "commonName": "",
    "accessionNumber": "",
    "mapLink": "",
    "imageUrl": "",
    "dnaSequences": [
      "",
      "",
      "",
      ""
    ]
  }
]
                    </pre>
                    </li>
                    <li STYLE="font-weight: normal">
                        You have already collected the scientific name, common name, and GenBank accession number
                        for each species, so add that data to the file.
                    </li>
                    <li STYLE="font-weight: normal">
                        To get the map link...
                        <ul>
                            <li>Go to <a href="https://www.google.com/maps">Google Maps</a> and click on the species' country of origin.</li>
                            <li>Click "Share", then "Embed a map", then "COPY HTML".</li>
                            <li>Paste inside the quotes after <code>"mapLink":</code> in "speciesData.json".</li>
                            <li>Delete the text surrounding the link (only the text inside the quotes after <code>src=</code> should remain).</li>
                            <li>Do this for each species.</li>
                        </ul>
                    </li>
                    <li STYLE="font-weight: normal">
                        To get the image URL...
                        <ul>
                            <li>Download JPEG photos of each species (keep track of image sources for copyright purposes).</li>
                            <li>If you desire, crop the photos to all have the same proportions.</li>
                            <li>Upload the photos to <a href="https://www.flickr.com/">Flickr</a> (or another image hosting website of your choice).</li>
                            <li>When a photo is selected (in Flickr), click the download button, then select "View all sizes".</li>
                            <li>Right click the image and select "Copy image address" to copy the static URL of the image.</li>
                            <li>Paste inside the quotes after <code>"imageURL":</code> in "speciesData.json".</li>
                            <li>Do this for each species.</li>
                        </ul>
                    </li>
                </ol>
            </li>

            <br/>

            <li STYLE="font-weight: bold">
                Get the DNA comparisons:
                <ol type="a">
                    <li STYLE="font-weight: normal">
                        Visit <a href="https://www.ebi.ac.uk/Tools/msa/muscle/">https://www.ebi.ac.uk/Tools/msa/muscle/</a>
                        to use the MUSCLE sequence alignment tool.
                    </li>
                    <li STYLE="font-weight: normal">
                        In "STEP 1 - Enter your input sequences", upload the file "speciesFASTA.txt".
                    </li>
                    <li STYLE="font-weight: normal">
                        Leave the default settings in "STEP 2 - Set your Parameters", then in "STEP 3 - Submit your job"
                        click "SUBMIT". (Depending on how many species are in your tree, this may take a while. You can
                        choose to be notified by email is you wish.)
                    </li>
                    <li STYLE="font-weight: normal">
                        When the result are ready, visually scan the alignment to find four consecutive segments with high
                        variation but preferably no insertions or deletions. Copy and paste the four lines of DNA for
                        each species in the array after <code>"dnaSequences":</code> in "speciesData.json".
                    </li>
                    <li STYLE="font-weight: normal">
                        Open "speciesComparisons.json" and paste the following (with one JSON object for each non-leaf
                        node in your tree):
                        <pre>

[
  {
    "key": "",
    "compStrings": [
      "",
      "",
      "",
      ""
    ]
  },
  {
    "key": "",
    "compStrings": [
      "",
      "",
      "",
      ""
    ]
  }
]
                        </pre>
                    </li>
                    <li STYLE="font-weight: normal">
                        The <code>"key"</code> must have the scientific names of each species that shares the common ancestor
                        represented by the non-leaf node. The scientific names need to be in alphabetical order, all lowercase,
                        and separated only by underscores.
                        <ul>
                            <li>
                                For example, imagine you have a tree containing three species of butterfly: Monarch Butterfly
                                (Danaus Plexippus), Silver-Studded Blue (Plebejus Argus), and Common Yellow Swallowtail
                                (Papilio Machaon). The root node represents the common ancestor of all three butterflies, so
                                the DNA comparison from that node will show the DNA alignment for all three butterflies. The
                                <code>"key"</code> for this comparison will be <code>"danaus_plexippus_papilio_machaon_plebejus_argus"</code>.
                            </li>
                        </ul>
                    </li>
                    <li STYLE="font-weight: normal">
                        Collect the comparison strings starting with the root node (the <code>"key"</code> should have the
                        scientific name of every species in your tree). Using the same four consecutive segments in
                        your MUSCLE alignment, copy and paste the four strings of asterisks and spaces in the array after
                        <code>"compStrings":</code>.
                    </li>
                    <li STYLE="font-weight: normal">
                        For each remaining <code>"key"</code>, you will need to get new comparison strings (of the same
                        four segments) for only the species who share that common ancestor. To do this, create a mini
                        FASTA file for each <code>"key"</code> using the DNA in "speciesData.json".
                        <ul>
                            <li>
                                In our butterfly example, pretend that the Monarch Butterfly and Common Yellow Swallowtail
                                are most closely related. For the node that joins these two species, the <code>"key"</code>
                                would be <code>"danaus_plexippus_papilio_machaon"</code>, and the mini FASTA file would look like this:
                                <pre>

>Monarch Butterfly|(Danaus Plexippus)
AATCCCTCAAGGAACTCCTTTTATTTTAATACCTTTTATAGTATTAATTGAAACAATTAG
AAATATTATTCGACCAGGCACTTTAGCTATTCGTTTAACAGCTAATATAATTACTGGTCA
TCTTTTAATCACTTTATTAAGAAGAATAGGAAACAGAATTAATTTTTATTTAATTTCACT
TTTAATTTTTATACAAATTTTATTATTAATTTTAGAATCTATAGTAGCAATTATTCAATC

>Common Yellow Swallowtail|(Papilio Machaon)
AATCCCGCAAGGAACTCCTTCAATTTTAATGCCTTTTATAGTAATAATTGAAACAATTAG
AAATATTATTCGTCCTGGAACATTAGCTGTTCGTCTATCTGCTAATATAATTGCCGGTCA
TTTATTATTAACTCTTCTAAGAAGAACTAGTAATAGTTTATCATTTTATTTATTATTTAT
TTTAATTTTTATACAAATTTTATTATTAATTTTAGAATCAGCAGTTGCAGTTATTCAAGC
                                </pre>
                            </li>
                            <li>
                                Running this alignment in MUSCLE will be very quick. Copy and paste the strings of asterisks and
                                spaces in the array after <code>"compStrings":</code>, then repeat for each remaining <code>"key"</code>.
                            </li>
                        </ul>
                    </li>
                </ol>
            </li>

            <br/>

            <li STYLE="font-weight: bold">
                ALL DONE!
                <ol type="a">
                    <li STYLE="font-weight: normal">
                        Drag and drop "speciesTree.json", "speciesData.json", and "speciesComparisons.json" into the
                        appropriate spots on the <a href="tree_drop.html">"Upload Your Own"</a> page of the QuickTree website.
                    </li>
                </ol>
            </li>

        </ol>
        <hr/>
        <br/>
        <br/>
    </div>
</body>
</html>